adapter/network_management.md

```diff
@@ -0,0 +1,253 @@
+| in   | cmd.network.node_update      | int        | Requests an interview and discovery of its capabilities for the provided address of a thing.                      |
+| in   | cmd.network.reset            | null       | Requests forceful removal of all things and resets the network.                                                   |
+| out  | evt.network.reset_done       | null       | Reports that network reset has been completed.                                                                    |
+
+### Definitions
+
+* `node_report` is an object with the following structure:
+
+| Field           | Type   | Example                  | Description                                                                                                                         |
+|-----------------|--------|--------------------------|-------------------------------------------------------------------------------------------------------------------------------------|
+| address         | string | `"4"`                    | Unique address of a thing, equal to `address` field in [inclusion_report](/adapter/thing_management.md#definitions) for this thing. |
+| hash            | string | `"zw_398_3_12"`          | Equal to `product_hash` field in [inclusion_report](/adapter/thing_management.md#definitions) for this thing.                       |
+| alias           | string | `"Fibaro Double Switch"` | Equal to `product_name` field in [inclusion_report](/adapter/thing_management.md#definitions) for this thing.                       |
+| power_source    | string | `"ac"`                   | Equal to `power_source` field in [inclusion_report](/adapter/thing_management.md#definitions) for this thing.                       |
+| wakeup_interval | string | `"3600"`                 | Equal to `wakeup_interval` field in [inclusion_report](/adapter/thing_management.md#definitions) for this thing.                    |
+| comm_tech       | string | `"zw"`                   | Equal to `comm_tech` field in [inclusion_report](/adapter/thing_management.md#definitions) for this thing.                          |
+| status          | string | `"UP"`                   | Status of a thing, either `UP` or `DOWN`, see [`node_status_matrix`](#definitions).                                                 |
+| sub_status      | string | `"online"`               | Sub status of a thing, one of `online`, `sleeping`, `updating`, `configuring`, `offline`, see [`node_status_matrix`](#definitions). |
+| conn_quality    | string | `"high"`                 | Network connection quality rating, one of `low`, `medium`, `high`, or `undefined` if unknown or not applicable.                     |
+| conn_type       | string | `"direct"`               | Either `direct` or `indirect`, if communication is going through hops or third party cloud services/proxies.                        |
+
+* `node_status_matrix` is a matrix of possible `status` and `sub_status` values:
```

Comment:
Issue 3: This is a proposal of a status/sub_status matrix. Feel free to propose changes or alternatives. The goal is to have in future a single and readily available place to be able to asses status of the thing, it's connection quality and readiness, which may be impacted by temporary process such as discovery, firmware update or sleeping.

What we need to consider are also UX and certification requirements, an application might need to show the user information that the device i sleeping and any issued command might be received and executed later. We will also want to show the type of connection of the device. It is important to make it easier for the mobile app to access and use this data, as well as other services running in background on the hub.

Reply:
<img width="1052" alt="image" src="https://user-images.githubusercontent.com/31538120/224343041-35f4cae2-4794-4b39-81cf-cd835576a214.png">

Reply:
It seems okay

Reply:
UP/DOWN I call it reachability in the code. We have bunch of 'statuses' everywhere so maybe lets use REACHABILITY as more outstanding name

Reply:
It does not seem good to me.
- Online and offline seems as duplication of UP/DOWN

What if device was being updated and now request started to time out ? We need to separate these and define what they mean.

First we need to define what each category is about and describe separately to ensure they are disjoint - only then this matrix makes sense.
- STATUS defines if the device will response when inquired. I call it REACHABILITY in zigbee-ad to differentiate from many different 'statuses' elsewhere so I propose the same name here. AVAILABILITY or RESPONSIVE is another option.
- SUB_STATUS hard to say for me what would it be as it's a mixture of different things. Device can be updating and sleeping in the meantime jsut after getting the block od image...
This maybe should be separated to "OTA stage" (updating, discovering) and reachability (online, offline, sleeping)

Hmmm maybe **REACHABILITY** could be set {UP, DOWN, SLEEPING} -> would make more sens from my perspective. Exactl definition to be defined but just to share the idea it would be:
**DOWN**  for:
- AC devices when not reachable as they surely should be responsie all time
- DC devices when no responsive wihin interval <wakeUpInterval, inf>

**UP**  for:
- AC, DC devices recently (recetly - TBDefined) responsive

**SLEEP**  for:
- DC devices which were responsive wihtin  <0, wakeUpInterval>


SUB_SATUS -> maybe rename as LifePhase (I use the name to mark device as Associated (when broadcast packet before inclusion caught), Joined/Included, Discovered, Updating, Discovering, Excluding)
-> This is the state during device which device can be when connected to the network and its important how it behaves depending what actually is happening. Eg: I set device LifePhase to Discovery to reject all cmd.meter.get_report untill whole cluster metering is not discovered & divisors downloaded.

Reply:
@BudzikM 
Thanks for your comments. It's exactly what you said just the other way round. Let me elaborate better and then we can focus on better wording.

Status is the most important binary information:

DOWN - means that the device is considered to be offline/disconnected/unreachable/etc. If it is AC or battery FLIR (non-sleeping) device it means it failed to respond to a number of retries recently and did not send anything on its own. If it is a sleeping device it means it failed to send anything past its sleeping interval usually with some safety margin and did not respond to ping or queued commands during the wakeup period.

UP - just invert the description of DOWN.

Let me put additional notes. Device that you're busy interacting with or is conducing some internal process is UP.  If you're discovering/configuring the device, it is UP, even though you cannot effectively use it or send commands to it. If the device is actively downloading OTA image, it is UP, even though it is not useable at the moment. If device is sleeping, but it is expected for it to sleep now, it is also UP. 

For backwards compatibility we cannot change the property name "status" and we cannot change the values "UP" and "DOWN".

SUBSTATUS is a more detailed information indicating not only if the device is online but also if it is ready to receive commands. SUBSTATUS contains in itself STATUS but for backwards compatibility we need to maintain both of them, so effectively SUBSTATUS repeats and extends what STATUS says but also adds more information.

The matrix shows only all logical combinations of status and substatus.

We can rename substatus and its values however we want but it won't change the number of enums or their logical relation. Maybe instead of online/offline ready/not-ready would be better? I don't know. I don't like not-ready because it is too weak semantically. Das ding ist kaput und reportieren "not-ready"???!!!

Reply:
Minor comments:
* DC does not mean battery powered in FIMP, it means DC source, and it was meant mostly to be used for inverters  and solar panels.
* We probably don't need to map 1:1 with internal statuses kept in the adapter, mostly because we're interested in broadcasting a status from the time of first inclusion report being sent to the time when exclusion report is being sent.

So statuses like associated, joined, discovered probably do not make much sense. But it makes sense to indicate "discovering" status when you have enough information to send initial inclusion report, but then still doing gathering information about the device.

Reply:
1. "What if device was being updated and now request started to time out "
Then just use "Online" status once again if it's online. More descriptive info should be within OTA service.

2. "Device can be being updated and sleeping in the meantime just after getting the block of image..." 
You mean, that the image uploading to device has been done, but installation did not proceeded yet?

@boguslaw-wojcik maybe we should point out a proper usage of those sub_statuses? Like.. for substatus "updating", app would know based on that, that it shall listen and act on ota service and show statuses from that service or...?

Reply:
You want to ban some combination of this 'matrix' ? 
Please give me examples and under what condition its possible of:
-device is UP and offline
-device is DOWN and online
-device is DOWN and sleeping
-device is DOWN and updating

Reply:
So according to you "SUBSTATUS is a more detailed information" and STATUS is less detailed information about the same then ? It needs to be precisely defined what is what otherwise it would be the source of confusion for FIMP users.